man/man1/runradex.1
.TH RUNRADEX 1NEMO "8 January 2012"
.SH NAME
runradex \- NEMO frontend to run the RADEX program
.SH SYNOPSIS
\fBrunradex\fP mol= outdir= [parameter=value]
.SH DESCRIPTION
RADEX is a statistical equilibrium radiative transfer code. This particular
version is a one-dimensional non-LTE radiative
transfer code, that uses the escape probability formulation assuming
an isothermal and homogeneous medium without large-scale velocity
fields. RADEX is comparable to the LVG method and provides a useful
tool in rapidly analyzing a large set of observational data providing
constraints on physical conditions, such as density and kinetic
temperature.
.PP
RADEX provides an alternative to the widely used rotation diagram
method which relies upon the availability of many optically thin
emission lines and is useful only in roughly constraining the
excitation temperature in addition to the column density. RADEX makes
no extrapolation of collisional rate coefficients outside the
temperature range specified in the molecular data file used as
input. Instead they are kept constant.
.PP
In addition to this frontend to the original fortran program, there is
also an online version available (see below).
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBmol=\fP
Input molecular data file e.g. hco+.dat, from $RADEXDAT. If the
filename starts with . or /, it will be taken as an absolute or
relative address, else a predefined location (such as $RADEXDAT)
will be used. No default.
.TP
\fBoutdir=\fP
Run directory within which all files will be created. No default.
.TP
\fBout=\fP
Output filename [output.txt]     
.TP
\fBfmin=\fP
The statistical equilibrium calculation takes all spectral lines of
the molecular model into account, but the output only lists lines at
frequencies between the given minimum and maximum frequency given here.
Allowed range is 0.1 - 1e7 GHz.  [50]    
.TP
\fBfmax=\fP
Maximum frequency in GHz. [500]    
.TP
\fBtkin=\fP
Kinetic temperature of the medium in K.  Allowed range
is 0.1 - 10,000. [20]    
.TP
\fBpartner=\fP
Collision partner name(s). Pick (m)any from: H2,p-H2,o-H2,e,H,He.  [H2]
.TP
\fBdensity=\fP
Number density (cm^-3) of collision partners (usually just H2). By default,
collisions with He are not taken into account. To approximate their
effect, multiply the H2 density by 1.2 for an He/H2=0.2 abundance. The
effects of mass and geometrical cross section between He and H2 cancel
out to first order.
One number required for each partner. [1e4]    
.TP
\fBtbg=\fP
Rayleigh-Jeans equivalent temperature of the background radiation
field, assumed to be of black body shape. Default is the cosmic
microwave background. Units are K. [2.73]    
.TP
\fBcdmol=\fP
Column density of the molecule to calculate line strengths for,
in cm^-2.  [1e13]   
.TP 20
\fBdeltav=\fP
Thermal + turbulent width (FWHM) of the lines, in km/s. Allowed
range is 1e-3 to 1e3. [1.0]   
.SH EXAMPLES
This frontend is particularly useful to run a large number of models, optionally
followed by some more analysis. For example, here is a set of models
run at a grid of temperature and densities:
.nf
    foreach tkin (15 20 25 30 35)
      foreach cdmol (1e12 3e12 1e13 3e13 1e14)
        runradex hco+.dat run_${tkin}_$cdmol tkin=$tkin cdmol=$cdmol
      end
    end
.fi
this will create a series of run directories (run_15_1e12,run_15_1e13, ... run_35_3e13)
with a local file called output.txt containing the output from RADEX. There are also
two python scripts in the distribution of RADEX with examples how to run over a grid
of models and extract or fit data.
.SH SEE ALSO
http://www.sron.rug.nl/~vdtak/radex/index.shtml
.PP
http://www.strw.leidenuniv.nl/~moldata/ - molecular data (and their format description)
.PP
Van der Tak, F.F.S., Black, J.H., Schöier, F.L., Jansen, D.J., van Dishoeck, E.F., 2007, A&A 468, 627-635.
.SH FILES
.nf
$NEMO/usr/radex/runradex.c - source code
$RADEXDAT/hco+.dat  - default example molecular data file
$rundir/input.txt   - input text file for radex itself, generated by runradex
$rundir/radex.log   - logfile , created by radex
.fi
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
07-Jan-12	V0.1 Created	PJT
.fi


usr/radex/runradex.c
/*
 *  RUNRADEX:   run the RADEX code
 *
 *    This program only assembles the parameter to write the input parameter
 *    file, copies your input file to the properly named qmics.dat file,
 *    stuffs all of that in a (new) run directory, and runs the program in
 *    that directory.
 *
 */

#include <nemo.h>

string defv[] = {
  "mol=???\n             Input molecular data file [hco+.dat, from $RADEXDAT/]",
  "outdir=???\n          Run directory",
  "out=output.txt\n      Output filename",
  "fmin=50\n             Minimum freq [GHz]",
  "fmax=500\n            Maximum freq [GHz]",
  "tkin=20\n             Kinetic temperature [K]",
  "partner=H2\n          Partner(s) names (H2 p-H2 o-H2 e H He H+)",
  "density=1e4\n         Density of partner(s)",
  "tbg=2.73\n            Background temperature [K]",
  "cdmol=1e13\n          Molecular column density [cm^-2]",
  "deltav=1.0\n          Line width [km/s] ?",
  "VERSION=0.1\n         7-jan-2012 PJT",
  NULL,
};

string usage="RADEX frontend; see also http://www.sron.rug.nl/~vdtak/radex/index.shtml";

string cvsid="$Id$";

#define MAXPARTNER   7

void nemo_main()
{
  int n;

  stream parstr, outstr;
  string molfile = getparam("mol");
  string outfile = getparam("out");
  string outdir  = getparam("outdir");
  string exefile = "radex";
  string datfile = "mol.dat";
  string parfile = "input.txt";
  string logfile = "radex.txt";
  char dname[256], cmd[256];

  real  fmin = getdparam("fmin");
  real  fmax = getdparam("fmax");
  real  tkin = getdparam("tkin");

  int   npartner;
  string *partner;
  real  density[MAXPARTNER];

  real  tbg = getdparam("tbg");
  real  cdmol = getdparam("cdmol");
  real  deltav = getdparam("deltav");

  partner = burststring(getparam("partner"),",");
  npartner = xstrlen(partner,sizeof(string))-1;
  if (npartner < 1 || npartner>MAXPARTNER) error("too few/many partners");
  n = nemoinpr(getparam("density"),density,MAXPARTNER);
  if (n != npartner) error("need matching numbers for partner= and density=");

  make_rundir(outdir);
  sprintf(dname,"%s/%s",outdir,parfile);

  parstr = stropen(dname,"w");

  fprintf(parstr,"%s\n",molfile);
  fprintf(parstr,"%s\n",outfile);
  fprintf(parstr,"%g %g\n",fmin,fmax);
  fprintf(parstr,"%g\n",tkin);
  fprintf(parstr,"%d\n",npartner);
  for (n=0; n<npartner; n++) {
    fprintf(parstr,"%s\n",partner[n]);
    fprintf(parstr,"%g\n",density[n]);
  } 
  fprintf(parstr,"%g\n",tbg);
  fprintf(parstr,"%g\n",cdmol);
  fprintf(parstr,"%g\n",deltav);
  fprintf(parstr,"0\n");

  strclose(parstr);

  goto_rundir(outdir);
  sprintf(cmd,"%s < %s > %s ", exefile, parfile, logfile);
  run_program(cmd);
}

goto_rundir(string name)
{
  dprintf(0,"CHDIR: %s\n",name);
  if (chdir(name))
    error("Cannot change directory to %s",name);
}

make_rundir(string name)
{
  dprintf(0,"MKDIR: %s\n",name);
  if (mkdir(name, 0755))
    warning("Run directory %s already exists",name);
}

run_program(string cmd)
{
  dprintf(0,"EXEC: %s\n",cmd);
  system(cmd);
}


